Fix/run multiple blocks

MaikePaetzel opened the pull request:
The robot is now able to perform multiple walks and turns in a row. The ball behavior still needs to be updated in a similar fashion, though.

schra committed:
[feat] highlight the block currently executed

schra committed:
[patch] Use acorn JS interpreter instead of eval()

Acorn JS interpreter: https://github.com/NeilFraser/JS-Interpreter

The Acorn JS interpreter is recommended for Blocky:
https://developers.google.com/blockly/guides/app-integration/running-javascript#js-interpreter

schra committed:
[wip] [fix] running multiple blocks

MaikePaetzel committed:
[Fix] Robot can now perform multiple actions one after another. Also made the turning dynamic.

schra commented:
Very nice :) I like your approach with `isActive`.

I found three problems:

* This example yields infinite rotation (target rotation seems to not be reached?)

  ![Screenshot from 2021-04-16 22-50-19](https://user-images.githubusercontent.com/23171985/115082635-82eb2200-9f06-11eb-9e80-16b2239f3c8a.png)

* This example yields a JS error:

  ![Screenshot from 2021-04-16 22-48-43](https://user-images.githubusercontent.com/23171985/115082638-84b4e580-9f06-11eb-8461-230f119b0c1a.png)

* Micha implemented that when a player "goes over" a ball, the ball comes with the player (but he didn't implement a smooth transition of the ball - the ball just jumped with the player). This doesn't work anymore.

schra commented on src/helper/useBlockly.js:
```diff
@@ -27,22 +30,74 @@ export function useBlockly() {
+                    return;
+                }
+                workspaceCodeInterpreter.step();
+                workspaceCodeInterpreter.step();
```
I'm unsure whether there is *always* a highlight function after each function in the generated code (e.g. in the case of loops). My approach was to call `.step` only once and divide the interval by two. This way we are sure to never step over non-highlighting block.

MaikePaetzel replied:
Wasn't actually necessary to do that, I took the second call out (was for testing and I forgot to remove it again)

MaikePaetzel committed:
[Fix] Left turn now works as well

MaikePaetzel committed:
[Fix] removed stepping over two blocks in a row (wasn't necessary, just for testing and forgot to take it out)

MaikePaetzel commented:
>     * This example yields infinite rotation (target rotation seems to not be reached?)

Fixed left rotation.

>     * This example yields a JS error:

Was actually a problem with the left rotation as well, so fixed as well.

>     * Micha implemented that when a player "goes over" a ball, the ball comes with the player (but he didn't implement a smooth transition of the ball - the ball just jumped with the player). This doesn't work anymore.

The entire ball interaction doesn't work yet (see my initial message in the PR). Someone else is welcome to take the example of how I implemented the behavior update as part of the Reducer and do the same changes for the ball :)

schra reviewed (changes requested):
I get a JS error with this code:

![Screenshot from 2021-04-16 23-47-39](https://user-images.githubusercontent.com/23171985/115087400-5e934380-9f0e-11eb-8a92-4b17021cb0e3.png)


```
Uncaught TypeError: rob is undefined
    nextStep useBlockly.js:82
    interpret useBlockly.js:88
    tick useInterval.js:20
useBlockly.js:82
    nextStep useBlockly.js:82
    interpret useBlockly.js:88
    tick useInterval.js:20
```

MaikePaetzel committed:
[Fix] ball behavior now updates correctly with multiple blocks as well. For simplification, combined the Ball and Robot reducer into one GameStateReducer

MaikePaetzel committed:
[Feature] refactored the task logic - now the task components are defined by what blocks are available and the exact configuration of ball and robot are set in task.json

MaikePaetzel committed:
[Fix] turn for more than 360 deg now correctly resets to 0

schra committed:
[fix] Removing blocks during execution

This is unused code anyway.

MaikePaetzel committed:
[Fix] left turn now works as well

schra committed:
[fix] remove need to call updateBall and moveBall in tasks

schra committed:
[patch] streamline naming of actions and reducers

schra committed:
[fix] make robot turn correctly. refactored code
